RavenDB-15927 Collection queries are reading their entire input befor…

ayende opened the pull request:
…e they start running

ayende committed:
RavenDB-15927 Collection queries are reading their entire input before they start running

ppekrol commented on src/Raven.Server/Documents/CollectionRunner.cs:
```diff
@@ -85,36 +85,42 @@ public virtual Task<IOperationResult> ExecuteDelete(string collectionName, long
+                                ids.Enqueue(document.Id);
+
+                                if (ids.Count >= OperationBatchSize)
+                                    break;
```
The whole problem is that the collection query is:
1. Not supporting `DocumentField` - should be easy to do?
2. Shouldn't it support paging there, like the others inside 'GetDocuments' ?

grisha-kotler replied:
We already request `OperationBatchSize` in the `foreach`

ayende replied:
Yeah, that wasn't handled for `CollectionQueryRunner`, fixing it.

grisha-kotler commented on src/Raven.Server/Documents/CollectionRunner.cs:
```diff
@@ -85,36 +85,42 @@ public virtual Task<IOperationResult> ExecuteDelete(string collectionName, long
                         foreach (var document in GetDocuments(context, collectionName, startEtag, internalQueryOperationStart, OperationBatchSize, isAllDocs, DocumentFields.Id))
                         {
-                            internalQueryOperationStart++;
+                            using (document)
```
We dispose here only the document id and we collect here `OperationBatchSize` which is 1024 IIRC.

ayende replied:
I want to ensure that if we have encrypted database, we return the buffers ASAP

ayende committed:
RavenDB-15927 Passing paging limit to the IndexQueryServerSide

ayende committed:
RavenDB-15927 Send DocumentFields.Id to the CollectionQueryEnumerable to reduce the amount of work we have to do to process the query

ayende committed:
RavenDB-15927 Sending the appropriate flag

ppekrol committed:
RavenDB-15927
- fixed NRE - we need to extract data when we are no doing all docs collection starts with
- disposing document immediately if it does not match

ppekrol commented:
@ayende I've made some adjustments, can you review?

ppekrol committed:
RavenDB-15927 do not substring Id for wildcard match if there is nothing to match to

ppekrol committed:
RavenDB-15927 cleanup

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -697,58 +696,84 @@ public static long ReadLastEtag(Transaction tx)
-            foreach (var doc in GetDocumentsStartingWith(context, idPrefix, matches, exclude, startAfterId, start, take))
+            var isAllDocs = collection == Constants.Documents.Collections.AllDocumentsCollection;
+            var requestedDataField = fields.HasFlag(DocumentFields.Data);
+            if (isAllDocs == false && requestedDataField == false)
```
why do we need this? to read the metadat?

ppekrol replied:
Yes

ayende reviewed (commented):
LGTM

arekpalinski commented:
LGTM
